api/v1alpha1/ratelimitfilter_types.go

```diff
@@ -0,0 +1,133 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+
+// RateLimitFilter allows the user to limit the number of incoming requests
+// to a predefined value based on attributes within the traffic flow.
+type RateLimitFilter struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of HTTPRoute.
+	Spec RateLimitFilterSpec `json:"spec"`
+}
+
+// +kubebuilder:object:root=true
+
+// RateLimitFilterList contains a list of RateLimitFilter resources.
+type RateLimitList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []RateLimitFilter `json:"items"`
+}
+
+// RateLimitFilterSpec defines the desired state of RateLimitFilter
+// +union
+type RateLimitFilterSpec struct {
+	// Type decides the scope for the RateLimits.
+	// +unionDiscriminator
+	Type RateLimitType `json:"type"`
+	// Rules are a list of RateLimit matchers and limits.
+	//
+	// +kubebuilder:validation:MaxItems=16
+	Rules []RateLimitRule `json:"rules"`
+}
+
+// RateLimitType specifies the types of RateLimiting.
+// Valid RateLimitType values are:
+//
+// * "Global"
+//
+// +kubebuilder:validation:Enum=Global
+type RateLimitType string
+
+const (
+	// In this mode, the rate limits are applied across all Envoy proxy instances.
+	RateLimitTypeGlobal RateLimitType = "Global"
+)
+
+// RateLimitRule defines the semantics for matching attributes
+// from the incoming requests, and setting limits for them.
+type RateLimitRule struct {
+	// +optional
+	// +kubebuilder:validation:MaxItems=8
+	Matches []RateLimitMatch `json:"matches,omitempty"`
```

Comment:
As @sunjayBhatia mentions [here](https://github.com/envoyproxy/gateway/pull/706#discussion_r1051165778), I prefer the API start without `Matches`. I provide additional context [here](https://github.com/envoyproxy/gateway/pull/706#discussion_r1052537026).